### PR TITLE
Extract config root from pwd instead of using scriptnames

### PR DESCRIPTION
**Pull request by leiserfg**

### CODE REVIEW

**Comment by L3MON4D3** [comment]
Makes sense👍
Could you apply this commit to the first PR and see whether tests still pass?
What do you think about doing a different kind of fallback? I mean, if the config is not sourced, there just is no CONFIG_ROOT. It might be more sensible to fallback to.. maybe just HOME

**Comment by leiserfg** [comment]
This allows to copy your full config to another folder (like  `.config/nvim/`  to `/tmp/nvim`)  do some changes and be able to use  `nvim -u /tmp/nvim/init.lua`

**Comment by leiserfg** [comment]
I rebased your branch on top of this one and the tests pass.

**Comment by L3MON4D3** [comment]
> This allows to copy your full config to another folder (like `.config/nvim/` to `/tmp/nvim`) do some changes and be able to use `nvim -u /tmp/nvim/init.lua`

Okay, yeah that makes some sense, but wouldn't unsourced `plugin`-files cause issues then?

**Comment by L3MON4D3** [comment]
> I rebased your branch on top of this one and the tests pass.

Ahh, great👍

**Comment by leiserfg** [comment]
> > This allows to copy your full config to another folder (like `.config/nvim/` to `/tmp/nvim`) do some changes and be able to use `nvim -u /tmp/nvim/init.lua`
> 
> Okay, yeah that makes some sense, but wouldn't unsourced `plugin`-files cause issues then?

Other plugins if they depend on the path (most o the time they don't), we do because we (potentially) read files relative to the config.

**Comment by L3MON4D3** [comment]
No, I mean if you have some stuff in your local `ftplugin`, for example. That wouldn't be sourced, right?
Like, is it an usual workflow to just copy everything?

The reason I'm bringing this up is because, beside `-u something`, `--clean` also leads to an empty `$MYVIMRC`, and I'm not sure those two are all cases, so handling each differently by checking commandline-arguments seems brittle.
(I think a simple "if not set then set to ..." might make more sense, since it's predictable what `./` means (maybe just current directory might be best, even), but if you have another solution, I'm all ears)

**Comment by leiserfg** [comment]
I'm just going for "keeping the previous behavior", because the only use case I know for checking a nonexisting `$MYCONFIG` is when testing configs. I don't think the `--clean` only case is important as in that case the plugin won't even load.

**Comment by L3MON4D3** [comment]
Keeping the existing behaviour makes sense, that's right..
But even in `--clean`, the plugin might be loaded manually in some way.

Any lua-config loaded via `-u lua_config.lua` is probably one crafted specifically for this occasion (MWE, probably). As such, we don't need to ensure compatibility with whatever setup exists otherwise (so it doesn't matter if a `load`-path has to be changed).

Setting `$MYCONFIG_ROOT` to the cwd seems the most sensible thing for MWEs, especially since then, we can just state in the docs that, if `$MYVIMRC` is not set, `$MYCONFIG_ROOT` resolves to cwd, which is precise(as opposed to manually handling commandline-arguments) and simple.

**Comment by leiserfg** [comment]
Then it is a breaking change, I don't think many will suffer with it but still one so it should be in [Breaking Changes](https://github.com/L3MON4D3/LuaSnip/issues/81). Fortunately, it was a non-documented feature.

**Comment by L3MON4D3** [comment]
> Then it is a breaking change, I don't think many will suffer with it but still one so it should be in [Breaking Changes](https://github.com/L3MON4D3/LuaSnip/issues/81). Fortunately, it was a non-documented feature.

no doc -> no breaking changes, ezpz :P

I'll add it there before merging :+1:

File: lua/luasnip/util/path.lua [hunk 1]
```diff
@@ -51,18 +51,15 @@ function Path.read_file(path)
 	return buf
 end
 
-local MYCONFIG_ROOT = vim.env.MYVIMRC
--- if MYVIMRC is not set then it means nvim was called with -u
--- therefore the first script is the configuration
--- in case of calling -u NONE the plugin won't be loaded so we don't
--- have to handle that
+local MYCONFIG_ROOT = vim.env.MYCONFIG_ROOT
```
Comment (L3MON4D3) [inline lua/luasnip/util/path.lua:54]: This is here so it can be overridden from the outside?

File: lua/luasnip/util/path.lua [hunk 2]
```diff
@@ -51,18 +51,15 @@ function Path.read_file(path)
...
 
 if not MYCONFIG_ROOT then
-	MYCONFIG_ROOT = vim.fn.execute("scriptnames"):match("1: ([^\n]+)")
+	if vim.env.MYVIMRC then
+		MYCONFIG_ROOT = vim.env.MYVIMRC:gsub(("%s[^%s]+$"):format(sep, sep), "")
```
Comment (L3MON4D3) [inline lua/luasnip/util/path.lua:58]: What do you think of using `fnamemodify` here?
Would make this a bit more readable

File: lua/luasnip/util/path.lua (hunk 1 above)
Comment (leiserfg) [reply lua/luasnip/util/path.lua:54]: Yes, if you don't like it I can remove it, I thought it was what you meant in your last comment.

**Comment by leiserfg** [comment]
> > Then it is a breaking change, I don't think many will suffer with it but still one so it should be in [Breaking Changes](https://github.com/L3MON4D3/LuaSnip/issues/81). Fortunately, it was a non-documented feature.
> 
> no doc -> no breaking changes, ezpz :P
> 
> I'll add it there before merging +1

https://xkcd.com/1172/

File: lua/luasnip/util/path.lua (hunk 1 above)
Comment (L3MON4D3) [reply lua/luasnip/util/path.lua:54]: Ohhh, no I didn't mean that. Though I can see I didn't word that all that well.
I don't think we need that mechanism.

**Comment by L3MON4D3** [comment]
> > > Then it is a breaking change, I don't think many will suffer with it but still one so it should be in [Breaking Changes](https://github.com/L3MON4D3/LuaSnip/issues/81). Fortunately, it was a non-documented feature.
> > 
> > 
> > no doc -> no breaking changes, ezpz :P
> > I'll add it there before merging +1
> 
> https://xkcd.com/1172/

Ahhh that's a good one xD

**Comment by L3MON4D3** [comment]
Okay, code's good, could you rebase so this is just one commit? (best include `Breaking Change` in the first line of the message).

**Comment by L3MON4D3** [comment]
Almost forgot about this one :sweat_smile:

**Pull request merged**